### PR TITLE
Extended offset table

### PR DESCRIPTION
**Pull request by erikogabrielsson**
Support for reading and writing extended table offset.

### CODE REVIEW

File: test/test_save.py [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+    @classmethod
+    def setUpClass(cls):
+        cls.tiled_size = Size(10, 10)
+        cls.frame_count = cls.tiled_size.area
+        MIN_FRAME_LENGTH = 2
+        MAX_FRAME_LENGTH = 100
+        # Generate test data by itemizing random bytes of random length
+        # from MIN_FRAME_LENGTH to MAX_FRAME_LENGTH.
+        cls.test_data = [
+            urandom(randint(
```
Comment (jesper-molin) [inline test/test_save.py:124]: You typically want your tests to be repeatable in case they fail. So I would remove the randomness here either by using a seed or something else.

File: wsidicom/instance.py [hunk 2]
```diff
@@ -543,8 +542,124 @@ def slice_thickness(self) -> float:
         """Return slice thickness."""
         return self._slice_thickness
 
+    def set_dataset_as_tiled_full(
+        self,
+        image_data: List[Tuple[Tuple[str, float], 'ImageData']]
+    ):
+        """Set dataset as a tiled full dataset. Make a new Shared functional
```
Comment (jesper-molin) [inline wsidicom/instance.py:549]: Could you clarify in the docstring what is happening here, are you converting a TILED_SPARSE to TILED_FULL, are setting the metadata to TILED_FULL because the image data is already tiled full. The wording 'Set dataset as a tiled full dataset" is a little bit unspecific.

File: wsidicom/instance.py [hunk 3]
```diff
@@ -556,20 +671,36 @@ def __init__(self, filepath: Path):
...
+        ) -> bool:
+            return tag in pixel_data_tags
+        dataset = read_partial(
+            cast(BinaryIO, self._fp),
+            _stop_at,
+            defer_size=None,
+            force=False,
+            specific_tags=None,
+        )
+        # dataset = pydicom.dcmread(self._fp, stop_before_pixels=True)
```
Comment (jesper-molin) [inline wsidicom/instance.py:703]: Remove or uncomment?

File: wsidicom/instance.py [hunk 4]
```diff
@@ -663,65 +782,118 @@ def get_filepointer(
...
+                f"BOT length should be a multiple of {EOT_BYTES}"
+            )
+        # Read the EOT into bytes
+        eot = self._fp.read(eot_length)
+        # Read EOT lengths
+        tag = self._fp.read_tag()
+        if tag != Tag('ExtendedOffsetTableLengths'):
+            raise WsiDicomFileError(
+                self.filepath,
+                f"Expected EOT lengths after EOT, found {tag}"
```
Comment (jesper-molin) [inline wsidicom/instance.py:839]: Make it a little more googlible: "Expected ExtendedOffsetTableLengths tag after EOT, found {tag}"

File: wsidicom/instance.py [hunk 5]
```diff
@@ -817,23 +961,67 @@ def read_frame(self, frame_index: int) -> bytes:
...
+        tag = self._fp.read_tag()  # Either EOT or Pixel data tag
+        frame_positions = []
+        table = None
+        table_type = 'bot'
+        if tag == Tag('ExtendedOffsetTable'):
+            table_type = 'eot'
+            table = self._read_eot()
+            tag = self._fp.read_tag()
+
+        if tag == Tag('PixelData'):
```
Comment (jesper-molin) [inline wsidicom/instance.py:983]: This if-tree is a little bit hard to follow, reorder to put special paths first and use binomial variable to document the the if-clauses if needed.

File: wsidicom/instance.py [hunk 6]
```diff
@@ -1082,13 +1285,115 @@ def get_encoded_tiles(
...
+        ----------
+        List[bytes]
+            Tiles in bytes.
+        """
         return [
             self._get_encoded_tile(tile, z, path) for tile in tiles
         ]
 
-    def get_encoded_scaled_tiles(
+    def get_scaled_tile(
```
Comment (jesper-molin) [inline wsidicom/instance.py:1309]: Now when people have started to use the library, renaming public functions is a bigger deal as you may break other peoples software. We have not promised a frozen API at this stage, but maybe we should document all renames in the release notes?

**Review by jesper-molin — COMMENTED** [review]
This adds functionality so this should be released to 0.2.0 right when you think it is time to release it?

File: test/test_save.py [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+                self.frame_count
+            ) as read_file:
+                tag = read_file._fp.read_tag()
+                self.assertEqual(tag, ItemTag)
+                length = read_file._read_tag_length(False)
+                BOT_ITEM_LENGTH = 4
+                self.assertEqual(length, BOT_ITEM_LENGTH*self.frame_count)
+                for frame in range(self.frame_count):
+                    self.assertEqual(read_file._fp.read_UL(), 0)
+                self.assert_end_of_file(read_file)
```
[the remark this answers is not in the document]
Comment (erikogabrielsson) [reply test/test_save.py:314]: Yes, it looks nicer.

File: wsidicom/instance.py [hunk 8]
```diff
@@ -257,9 +256,9 @@ def is_wsi_dicom(cls, datset: Dataset) -> bool:
             for module in module_attributes:
                 for attribute in module:
                     if attribute not in datset:
-                        warnings.warn(
-                            f' is missing {key} attribute {attribute}'
-                        )
+                        # warnings.warn(
+                        #     f' is missing {key} attribute {attribute}'
```
[the remark this answers is not in the document]
Comment (erikogabrielsson) [reply wsidicom/instance.py:260]: Removed

File: wsidicom/instance.py (hunk 3 above)
Comment (erikogabrielsson) [reply wsidicom/instance.py:703]: Remove

File: wsidicom/instance.py (hunk 6 above)
Comment (erikogabrielsson) [reply wsidicom/instance.py:1346]: Noted
